core/src/main/scala/cats/data/Chain.scala

```diff
@@ -760,19 +764,57 @@ sealed abstract class Chain[+A] extends ChainCompat[A] {
 
   /**
    * Remove duplicates. Duplicates are checked using `Order[_]` instance.
+   * 
+   * Example:
+   * {{{
+   * scala> import cats.data.Chain
+   * scala> val chain = Chain(1, 2, 2, 3)
+   * scala> chain.distinct
+   * res0: cats.data.Chain[Int] = Chain(1, 2, 3)
+   * }}}
    */
   def distinct[AA >: A](implicit O: Order[AA]): Chain[AA] = {
-    implicit val ord: Ordering[AA] = O.toOrdering
-
-    var alreadyIn = TreeSet.empty[AA]
+    if (isEmptyOrSingleton) this
+    else {
+      implicit val ord: Ordering[AA] = O.toOrdering
+
+      var result = Chain.empty[A]
+      val seen = mutable.TreeSet.empty[AA]
+      val it = iterator
+      while (it.hasNext) {
+        val next = it.next()
+        if (seen.add(next))
+          result = result :+ next
+      }
+      result
+    }
+  }
 
-    foldLeft(Chain.empty[AA]) { (elementsSoFar, b) =>
-      if (alreadyIn.contains(b)) {
-        elementsSoFar
-      } else {
-        alreadyIn += b
-        elementsSoFar :+ b
+  /**
+   * Remove duplicates by a predicate. Duplicates are checked using `Order[_]` instance.
+   * 
+   * Example:
+   * {{{
+   * scala> import cats.data.Chain
+   * scala> val chain = Chain(1, 2, 3, 4)
+   * scala> chain.distinctBy(_ / 2)
+   * res0: cats.data.Chain[Int] = Chain(1, 2, 4)
+   * }}}
+   */
+  def distinctBy[B](f: A => B)(implicit O: Order[B]): Chain[A] = {
+    if (isEmptyOrSingleton) this
+    else {
+      implicit val ord: Ordering[B] = O.toOrdering
+
+      var result = Chain.empty[A]
+      val seen = mutable.TreeSet.empty[B]
+      val it = iterator
+      while (it.hasNext) {
```

Comment:
one issue: this is going to create a `Chain` the costs `O(N)` to uncons.

I wonder if the better approach would build a Vector and `Wrap` that. Vector seems to be the default collection we use in other places

e.g.
```scala
val bldr = Vector.newBuilder[A]
...

Wrap(bldr.result())
```

Reply:
Yeah, I wonder too. I didn't really think about it, perhaps you're right. Let's try this option.

Reply:
fixed